src/EditorFeatures/Core/Snippets/RoslynLSPSnippetExpander.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Reflection;
+using System.Text;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    [Export(typeof(IRoslynLSPSnippetExpander))]
+    [Export(typeof(RoslynLSPSnippetExpander))]
+    internal class RoslynLSPSnippetExpander : IRoslynLSPSnippetExpander
+    {
+        protected object? _lspSnippetExpander;
+        protected Type? _expanderType;
+        protected MethodInfo? _expanderMethodInfo;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RoslynLSPSnippetExpander(
+            [Import("Microsoft.VisualStudio.LanguageServer.Client.Snippets.LanguageServerSnippetExpander", AllowDefault = true)] object? languageServerSnippetExpander)
+        {
+            _lspSnippetExpander = languageServerSnippetExpander;
+
+            if (_lspSnippetExpander is not null)
+            {
+                _expanderType = _lspSnippetExpander.GetType();
+                _expanderMethodInfo = _expanderType.GetMethod("TryExpand");
+            }
+        }
+
+        public bool TryExpand(TextChange textChange, SourceText sourceText, string? lspSnippetText, ITextView textView, ITextSnapshot textSnapshot)
+        {
+            if (_expanderMethodInfo is null)
+            {
+                return false;
+            }
```

Comment:
```suggestion
            Contract.ThrowIfFalse(CanExpandSnippet());
```

we shouldn't be calling into this if the caller already checked that it was legal to do snippets.

Reply:
Curious: What does Contract.XYZ in Roslyn do that's better than normal if is null etc. checks?

Reply:
`Contract` "blows up", meaning we get an actual failure (and watson etc.) so that we know we did something wrong and have to go fix it.   A null check like this masks a logic bug in our code, which can then manifest as broken behavior for users thatw e have no insight on what actually went wrong.  

In general, if something indicates a logic bug in our code, we do not try to be resilient to it.  Rather, we fail early and report the NFW so we can actionably address teh scenario.  This approach is born out by enumerable experiences trying to go the "be resilient and just trudge along" route, and how much extra cost that has given us over the years as we've attempted to track down failures and intermittent behavior for users.

Reply:
I c i c, ya I can absolutely see the value in capturing those NFW's to ensure we can investigate what went wrong properly. Thanks for clarifying!